### PR TITLE
Added Course schema and CRUD APIs on Course #30 

### LINKED ISSUE #30: feature request : CURD on courses
I want to create routes and services to provide CURD functionalities on courses.
- Create a new course
- Update the details of an existing course
- Delete a course
- Get details of a particular course

I will be creating the Schemas according to the requirements of the current frontend.

### PR DESCRIPTION
**Pull request by Mohit-Swain**
Things I did in this PR:
- created a mongoose schema of course
- API to get a course by id
- API to create a new course
- API to update a course
- API to delete a course

resolves #30

### CODE REVIEW

**Review by SoulNinja-dev — APPROVED** [review]
Yes looks good to me.

**Comment by SoulNinja-dev** [comment]
@YashKumarVerma can you also review and then merge it 👍

File: src/modules/course/course.controller.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      throw new InternalServerException();
+    }
+  }
+
+  /**
+   * @param {object} obj - An object having all required course attributes.
+   */
+  static async createCourse(obj) {
+    logger.info('[course]: create new course');
+    let { title, instructors, enrollments, featured, tags, playlistId } = obj;
```
Comment (YashKumarVerma) [inline src/modules/course/course.controller.js:34]: can we use const here in place of let? are these values being changed anywhere?

File: src/modules/course/course.controller.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    try {
+      const course = await CourseService.updateCourse(
+        id,
+        title,
+        instructors,
+        enrollments,
+        featured,
+        tags,
+        playlistId,
+      );
```
Comment (YashKumarVerma) [inline src/modules/course/course.controller.js:71]: passing an object is better than passing so many params

File: src/modules/course/course.controller.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    logger.info('[course]: create new course');
+    let { title, instructors, enrollments, featured, tags, playlistId } = obj;
+    try {
+      const course = await CourseService.createNewCourse(
+        title,
+        instructors,
+        enrollments,
+        featured,
+        tags,
+        playlistId,
```
Comment (YashKumarVerma) [inline src/modules/course/course.controller.js:42]: pass an object in place of so many params, then we do not need to care about their order and default values

File: src/modules/course/course.controller.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  /**
+   * @param {ObjectId} id - ObjectId of the course to get
+   */
+  static async getCourseById(id) {
+    if (!id) {
+      throw new BadRequestException();
+    }
+    logger.info('[course]: Get course no. ' + id);
```
Comment (YashKumarVerma) [inline src/modules/course/course.controller.js:87]: prefer lowercase in logs

File: src/modules/course/course.service.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    return CourseList;
+  }
+
+  /**
+   * Fetch a course by Id
+   * @param {ObjectId} id - ObjectId of the course to get
+   */
+  static async findCourseById(id) {
+    const course = await Course.findById(id);
+    if (!course) {
```
Comment (YashKumarVerma) [inline src/modules/course/course.service.js:52]: better to use comparisons with null or undefined in place of `!variable`

File: src/modules/course/course.service.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+   * @param {string} playlistId - YouTube playlist id
+   */
+  static async updateCourse(
+    id,
+    title,
+    instructors,
+    enrollments,
+    featured,
+    tags,
+    playlistId,
```
Comment (YashKumarVerma) [inline src/modules/course/course.service.js:105]: changing this to an object will reduce the number of params

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]

**Comment by Mohit-Swain** [comment]
@YashKumarVerma I have made the requested changes, please have a look at it.

File: src/modules/course/course.controller.js (hunk 4 above)
Comment (YashKumarVerma) [reply src/modules/course/course.controller.js:70]: ?

File: src/modules/course/course.service.js (hunk 5 above)
Comment (YashKumarVerma) [reply src/modules/course/course.service.js:52]: ??

File: src/modules/course/course.controller.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    logger.info('[course]: delete course no. ' + id);
+    try {
+      const course = await CourseService.deleteCourse(id);
+
+      return course;
+    } catch (e) {
+      logger.error('[course]: ' + e);
+      throw new InternalServerException();
+    }
```
Comment (YashKumarVerma) [inline src/modules/course/course.controller.js:96]: the logging line should be inside the try block, after calling the service. currently it runs regardless whether the operation was successful or not

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]

**Comment by Mohit-Swain** [comment]
@YashKumarVerma I made the changes.

File: src/modules/course/course.service.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+   * @param {object} courseObj - Object containing all the course attributes.
+   * @param {string} courseObj.title - The title of the course.
+   * @param {Array<{name : string}>} courseObj.instructors - The instuctors of the course.
+   * @param {Array<{user: objectId,rating: Number}>} courseObj.enrollments - The students enrolled in the course.
+   * @param {bool} courseObj.featured - Is the course featured
+   * @param {Array<string>} courseObj.tags - Tags related to the course
+   * @param {string} courseObj.playlistId - YouTube playlist id
+   */
+  static async createNewCourse(courseObj) {
+    const course_rating = this.getAvgRating(courseObj.enrollments);
```
Comment (YashKumarVerma) [inline src/modules/course/course.service.js:70]: use camilCase only

File: src/modules/course/course.service.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    const course = new Course({
+      ...courseObj,
+      course_rating,
```
Comment (YashKumarVerma) [inline src/modules/course/course.service.js:73]: camelCase

File: src/modules/course/course.service.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+   * @param {Array<string>} courseObj.tags - Tags related to the course
+   * @param {string} courseObj.playlistId - YouTube playlist id
+   */
+  static async updateCourse(id, courseObj) {
+    const course_rating = this.getAvgRating(courseObj.enrollments);
+    const savedCourse = await Course.findByIdAndUpdate(
+      id,
+      {
+        ...courseObj,
+        course_rating,
```
Comment (YashKumarVerma) [inline src/modules/course/course.service.js:95]: case

File: src/modules/course/course.service.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    const len = enrollmentsWithRatings.length;
+    if (len === 0) {
+      return undefined;
+    }
+    const sumofRatings = enrollmentsWithRatings.reduce((sum, enrollments) => {
+      return sum + enrollments.rating;
+    }, 0);
+
+    return sumofRatings / len;
+  }
```
Comment (YashKumarVerma) [inline src/modules/course/course.service.js:35]: what would be the usecase when we'd need to fetch only the rating count? Would it not be better if we get the rating along with other course details?

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]

File: src/modules/course/course.service.js (hunk 11 above)
Comment (Mohit-Swain) [reply src/modules/course/course.service.js:35]: It is a util function that will calculate the avg rating of all the enrolled students and then store it in the database. It doesn't get called by the controller, but it gets called by the Create course and update course functions to calculate the avg rating.

**Comment by YashKumarVerma** [comment]
kindly checkout : https://github.com/commclassroom/commclassroomBackend/issues/38

**Comment by YashKumarVerma** [comment]
any updates?

**Comment by Mohit-Swain** [comment]
Sorry, I was busy in the prev week. I am starting to continue from today

**Comment by Mohit-Swain** [comment]
@YashKumarVerma can you check the commits.

**Review by YashKumarVerma — APPROVED** [review]

**Comment by YashKumarVerma** [comment]
thanks !

**Pull request merged**